### PR TITLE
miral::ConfigFile

### PR DESCRIPTION
**Pull request by AlanGriffiths**
A utility class to locate and (re)load configuration files when they change. The actual `Loader` processing of the file is one customization point, the reloading strategy another.

### CODE REVIEW

**Review by tarek-y-ismail — APPROVED** [review]

**Comment by AlanGriffiths** [comment]
Split this out of #3531 as it is generally useful

**Comment by tarek-y-ismail** [comment]
Might also handle #3478

**Review by Saviq — CHANGES_REQUESTED** [review]
There's one thing that we could consider here: monitoring _all_ the applicable folders for new files. I mean, if a file shows isn't in `$XDG_CONFIG_HOME` on startup, but is added later, we won't consider it.

Also the `XDG_CONFIG_DIRS` handling is static - we only load from it once.

---

Could we have tests, please?

**Comment by Saviq** [comment]
Should we consider `$PWD`, too?

It's surprising that this: `--display-config=static=foo.display` doesn't look in `$PWD` for the config file.

**Comment by AlanGriffiths** [comment]
> Should we consider `$PWD`, too?

I vote not. The XDG Base Directory spec. doesn't mention `$PWD` (which can also change during the life of the application). I would be more inclined to drop support for absolute paths and treat relative paths as relative to the base directory list. So that servers could organise configuration files in a subdirectory. (I adopted what we currently do for .display, which may not be ideal.)

> It's surprising that this: `--display-config=static=foo.display` doesn't look in `$PWD` for the config file.

For arbitrary files I would agree. But _config_ files typically don't load from the current directory.

> There's one thing that we could consider here: monitoring _all_ the applicable folders for new files. I mean, if a file shows isn't in `$XDG_CONFIG_HOME` on startup, but is added later, we won't consider it.

Yes we do.

> Also the `XDG_CONFIG_DIRS` handling is static - we only load from it once.

We could, but is there any benefit to monitoring for system-wide configuration? Session wide configuration changes should be in `$XDG_CONFIG_HOME`.

> Could we have tests, please?

We could. Just seemed a bit much for glue code that hits the filesystem

**Comment by AlanGriffiths** [comment]
> Adding some paint to the bikeshed :) But `--display-config=static=foo.display` is kind of odd to me

Before we paint the wrong bikeshed, this is about configuration files used by a compositor, not files specified on the command-line (or in the configuration).

File: src/miral/reloading_config_file.cpp [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    std::filesystem::path const filename;
+    std::optional<std::filesystem::path> const directory;
+    std::optional<mir::Fd> const directory_watch_fd;
+
+    void register_handler(MirRunner& runner);
+    std::unique_ptr<miral::FdHandle> fd_handle;
+};
+
+miral::ReloadingConfigFile::Self::Self(MirRunner& runner, path file, Loader load_config) :
+    inotify_fd{inotify_init()},
```
Comment (RAOF) [inline src/miral/reloading_config_file.cpp:86]: We probably want to be using `inotify_init1(IN_CLOEXEC)` here, no? Make sure the watch isn't leaked to child processes.

**Review by RAOF — COMMENTED** [review]
> We could, but is there any benefit to monitoring for system-wide configuration? Session wide configuration changes should be in `$XDG_CONFIG_HOME`.

I think we should be consistent between system-wide and user-local configuration behaviour.

Also, for the `ubuntu-frame` configuration do we try to stick it in the user-local or system-wide location?

**Comment by AlanGriffiths** [comment]
OK, the following questions raised above have not been resolved:

1. How should we treat relative paths? They could be relative to the XDG base locations. (This would make monitoring more complex as we need to watch for subdirectory creation and for files within that.)
2. How do we treat absolute paths?
3. Should we monitor the `XDG_CONFIG_DIRS` paths? (This would make monitoring more complex as we need to prioritise one path over another.)

Also, not mentioned (yet):

4. Should we detect file deletion and load a config from a later base location?

The proposed code answers as follows:

1. How should we treat relative paths? Relative to `$PWD`, and monitor for changes
2. How do we treat absolute paths? As absolute paths, and monitor for changes
3. Should we monitor the `XDG_CONFIG_DIRS` paths? No
4. Should we detect file deletion and load a config from a later base location? No

Also, mentioned above, but I think resolved:

5. Should we look in `$PWD`? No

I think this is useful and understandable

File: src/miral/reloading_config_file.cpp [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+                try
+                {
+                    if (event.name == filename)
+                    {
+                        auto const& file = directory.value() / filename;
+
+                        if (std::ifstream config_file{file})
+                        {
+                            load_config(config_file, file);
+                        }
```
Comment (Saviq) [inline src/miral/reloading_config_file.cpp:171]: Else?

File: tests/miral/reloading_config_file.cpp [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+char const* const home = "/tmp/test_reloading_config_file/home";
+char const* const home_config = "/tmp/test_reloading_config_file/home/.config";
+char const* const xdg_conf_home = "/tmp/test_reloading_config_file/xdg_conf_dir_home";
+char const* const xdg_conf_dir0 = "/tmp/test_reloading_config_file/xdg_conf_dir_zero";
+char const* const xdg_conf_dir1 = "/tmp/test_reloading_config_file/xdg_conf_dir_one";
+char const* const xdg_conf_dir2 = "/tmp/test_reloading_config_file/xdg_conf_dir_two";
+
+TestReloadingConfigFile::TestReloadingConfigFile()
+{
+    std::filesystem::remove_all("/tmp/test_reloading_config_file/");
```
Comment (Saviq) [inline tests/miral/reloading_config_file.cpp:106]: I'd rather this be a generated temp directory, but I've been down this rabbit hole a bit recently…

We are leaking it, though?

**Review by Saviq — APPROVED** [review]
>     1. How should we treat relative paths? Relative to `$PWD`, and monitor for changes
>     2. How do we treat absolute paths? As absolute paths, and monitor for changes

I think I'm OK with this. But maybe we want to be more verbose about it? Logging which file we've loaded and/or are monitoring?

>     3. Should we monitor the `XDG_CONFIG_DIRS` paths? No

Sounds like a potential extension later.

It does relate to:

> > There's one thing that we could consider here: monitoring _all_ the applicable folders for new files. I mean, if a file shows isn't in `$XDG_CONFIG_HOME` on startup, but is added later, we won't consider it.
> 
> Yes we do.

OK I miswrote. `XDG_CONFIG_HOME` (or rather, `config_directory()`) is the only folder we're monitoring. If things happen within "earlier" `XDG_CONFIG_DIRS`, we won't notice. But that may be extreme.

>     4. Should we detect file deletion and load a config from a later base location? No

My only problem with that is that we're behaving differently live than on startup.

>     5. Should we look in `$PWD`? No

:+1:

---

Thanks for the tests! Happy they helped :)

File: tests/miral/reloading_config_file.cpp (hunk 3 above)
Comment (AlanGriffiths) [reply tests/miral/reloading_config_file.cpp:106]: We are leaking, yes. That is intentional: when tests fail it is far easier to examine what happened. (And it is small, on a temporary filesystem)

**Comment by AlanGriffiths** [comment]
> OK I miswrote. `XDG_CONFIG_HOME` (or rather, `config_directory()`) is the only folder we're monitoring. If things happen within "earlier" `XDG_CONFIG_DIRS`, we won't notice. But that may be extreme.

There are no "earlier" `XDG_CONFIG_DIRS`: `$XDG_CONFIG_HOME`/`$HOME/.config` is the first place considered

File: src/miral/reloading_config_file.cpp (hunk 2 above)
Comment (AlanGriffiths) [reply src/miral/reloading_config_file.cpp:171]: Else we can't open the file, so no point trying to load a config from it

**Comment by Saviq** [comment]
> There are no "earlier" `XDG_CONFIG_DIRS`: `$XDG_CONFIG_HOME`/`$HOME/.config` is the first place considered

There are, if there's more than one in `XDG_CONFIG_DIRS`, and only the last one is where you find the file. Either way, we're not monitoring _any_ `XDG_CONFIG_DIRS`, only looking at them at startup. That's something we should at least spell out. Unless I'm reading the code completely wrong, there's only ever one folder we're monitoring. And that's never any of `XDG_CONFIG_DIRS`.

**Comment by AlanGriffiths** [comment]
> > There are no "earlier" `XDG_CONFIG_DIRS`: `$XDG_CONFIG_HOME`/`$HOME/.config` is the first place considered
> 
> There are, if there's more than one in `XDG_CONFIG_DIRS`, and only the last one is where you find the file. Either way, we're not monitoring _any_ `XDG_CONFIG_DIRS`, only looking at them at startup. That's something we should at least spell out. Unless I'm reading the code completely wrong, there's only ever one folder we're monitoring. And that's never any of `XDG_CONFIG_DIRS`.

We're in agreement on what happens, just not clear on the referent to "earlier"

**Comment by Saviq** [comment]
> We're in agreement on what happens, just not clear on the referent to "earlier"

Anyway! I don't think it's a blocker, but we probably should spell it out that `XDG_CONFIG_DIRS` are only considered on startup.

**Review by Saviq — APPROVED** [review]

File: src/miral/reloading_config_file.cpp (hunk 2 above)
Comment (Saviq) [reply src/miral/reloading_config_file.cpp:171]: Let's log it, then?

**Review by Saviq — APPROVED** [review]
I would just like to see more logging here. But you have my ACK.

"Monitoring $path/$file for configuration changes"
"Loaded configuration from $path/$file"
"Failed to open $path/$file"

**Comment by RAOF** [comment]
Oh, no. While pondering
> Should we monitor the XDG_CONFIG_DIRS paths? (This would make monitoring more complex as we need to prioritise one path over another.)
I was *going* to answer “Yes, we should monitor `XDG_CONFIG_DIRS`”. But then I considered again, and this is kinda unusual behaviour, and might trip people up (eg: saving partial config changes and having things break).

I *do* think that we should support changing the system-wide config at runtime, and the standard way of doing that is something like “reload config on `SIGHUP`”. And if we implemented *that* for system-wide config, would we still want to monitor the user-local file for changes?

**Comment by Saviq** [comment]
> Oh, no. While pondering

To recap what we discussed:
- indeed maybe SIGHUP would be more appropriate to load the main `.config` file
  - it would be more in line with other long-running services
  - would allow for `systemctl --user reload …` when in a systemd session
  - avoids accidental / in-progress / incomplete saves
  - can be easily wrapped to cause live reloads
- display config we can keep reloading live, it's mostly used by Frame anyway - and so can use this util

**Comment by AlanGriffiths** [comment]
A (different, in addition) summary of discussion:

* This PR was never intended as the foundation of future dynamic configuration support
  - It just generalises what we do for .display (and could also support Frame's diagnostic) and makes it available
* We might generalise this to offer options: e.g. load-on-startup, auto-reload and reload-on SIGHUP

Maybe I should rename the class `ConfigFile` and add a strategy enum with `load-on-startup and auto-reload` options?

File: tests/miral/CMakeLists.txt [hunk 4]
```diff
@@ -69,7 +69,7 @@ target_link_libraries(miral-test-internal
 
 mir_add_wrapped_executable(miral-test NOINSTALL
     external_client.cpp
-    reloading_config_file.cpp
+        config_file.cpp
```
Comment (Saviq) [inline tests/miral/CMakeLists.txt:72]:
```suggestion
    config_file.cpp
```

**Review by Saviq — APPROVED** [review]
Whitespace nit, otherwise good with me.

**Pull request merged**